api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -472,7 +494,15 @@ public NestedField asRequired() {
       if (!isOptional) {
         return this;
       }
-      return new NestedField(false, id, name, type, doc);
+      return new NestedField(false, id, name, type, defaultValue, doc);
+    }
+
+    public boolean hasDefaultValue() {
+      return defaultValue != null;
+    }
+
+    public Object getDefaultValue() {
+      return defaultValue;
```

Comment:
I think dealing with null-default as no-default will work, but from an Avro schema point of view, there is a distinction. Do we want to maintain that distinction, or say that default values in Iceberg are non-null default values only? I am okay with the latter as long as we do not end up doing complex checks to disambiguate between null-default and no-default. This will happen if we do not need to distinguish anywhere to begin with.

Reply:
Also, I am okay with proceeding with this for now and re-evaluating later, but it is something to think about.

Reply:
Thanks for the review @wmoustafa! 
null-default values today are mapped to optional NestedField (with no default). This shall remain the same, given that for optional NestedFields with no defaults, if the field is not manifested then null will be used. In other words, the mapping will be as follows (which shall become clearer in next PR, which handles the schema mapping/conversion):
```
avro field                 ->  iceberg nested field    | what happens when field is not manifested 
------------------------------------------------------------------------------------------------
optional, no default       ->  optional, no default.   | null is used
required, no default       ->  required, no default.   | error
optional, non-null default ->  optional, with default  | default value is used
required, non-null default ->  required, with default  | default value is used
optional, null default     ->  optional, no default    | null is used
required, null default     ->  optional, no default    | null (same as avro's default value) value is used
```

Reply:
I had the same table in my scratch notes to reason about this :) Yes, agreed it should work as I mentioned, just discussing this from a semantic point of view and code complexity down the line, but seems the latter will not be an issue, and the former should not be an issue as long as the latter is not.